File: python/deleteAtend.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
    
class LinkedList:
    def __init__(self):
        self.head = None
    
    def insert_node(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        last_node = self.head
        while last_node.next:
            last_node = last_node.next

        last_node.next = new_node
    
    def delete_at_end(self):
        if self.head is None:
            print("list is empty:")
            return
        if self.head.next is None:
            self.head = None
            return
        second_node = self.head
        while second_node.next.next:
            second_node = second_node.next
        
        second_node.next = None

    def print_list(self):
        temp = self.head
        while temp:
            print(temp.data, end=" -> ")
            temp = temp.next
        print("None:")


def main():
    link_list=LinkedList()
    link_list.insert_node(1)
    link_list.insert_node(2)
    link_list.insert_node(3)
    print("Before deletion:")
    link_list.print_list()
    link_list.delete_at_end()
    print("After deletion:")
    link_list.print_list()


if __name__=="__main__":
    main()

File: javaFile/delete_at_end.java
class Node {
    int data;
    Node next;
    Node(int data) {
        this.data = data;
        this.next = null;

    }
}
class LinkedList {
    Node head;
    LinkedList() {
        this.head = null;
    }

    void insert_node(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            return;
        }
        Node current = head;
        while (current.next != null) {
            current = current.next;
        }
        current.next = newNode;
    }

    void delete_at_end() {
        if (head == null) {
            System.out.println("List is empty. Nothing to delete.");
            return;
        }
        if (head.next == null) {
            head = null;
            return;
        }
        Node second_last_node = head;
        while (second_last_node.next.next != null) {
            second_last_node = second_last_node.next;
        }
        second_last_node.next = null;
    }

    void printList() {
        Node current = head;
        while (current != null) {
            System.out.print(current.data + " -> ");
            current = current.next;
        }
        System.out.println("null");
    }

}

public class delete_at_end {
    public static void main(String[] args) {
        LinkedList list = new LinkedList();
        list.insert_node(10);
        list.insert_node(20);
        list.insert_node(30);
        list.insert_node(40);
        list.insert_node(50);
        System.out.println("Original List:");
        list.printList();
        list.delete_at_end();
        System.out.println("List after deleting at end:");
    }
}
